### PR TITLE
BUG (string): ArrowStringArray.find corner cases

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

**Comment by jorisvandenbossche** [comment]
This will only work if we also backport https://github.com/pandas-dev/pandas/pull/56792

File: pandas/tests/strings/test_find_replace.py [hunk 1]
```diff
@@ -939,13 +940,22 @@ def test_find_bad_arg_raises(any_string_dtype):
...
     )
+    if (
+        pa_version_under13p0
+        and isinstance(ser.dtype, pd.StringDtype)
+        and ser.dtype.storage == "pyarrow"
+    ):
+        # https://github.com/apache/arrow/issues/36311
+        mark = pytest.mark.xfail(reason="https://github.com/apache/arrow/issues/36311")
+        # raises pa.lib.ArrowInvalid with Negative buffer resize
+        request.node.add_marker(mark)
```
Comment (jorisvandenbossche) [inline pandas/tests/strings/test_find_replace.py:958]: Alternative could be to check the pyarrow version in the implementation, and in that case for older pyarrow use the python implementation? 

(I don't know whether this happens in general or only for very specific cases. Based on looking at the test, this is buggy whenever there are missing values? In theory could also check for that, and only then fallback)

**Comment by jorisvandenbossche** [comment]
> This will only work if we also backport #56792

Backporting that specific PR is of course certainly possible, but if we want to avoid in general to have to backport older fixes, we could also copy the `_str_find` implementation of ArrowEA to ArrowStringArray (i.e. actually duplicating it), and only do the actual de-duplication later.

**Comment by jorisvandenbossche** [comment]
(moving the most up to date implementation directly to the shared mixin (or to ArrowStringArray, depending on questions in https://github.com/pandas-dev/pandas/pull/59555), would actually also avoid requiring to backport the older PR)

**Comment by jbrockmendel** [comment]
@jorisvandenbossche ive lost track of all your comments across several places.  pls advise on the preferred course of action.

File: pandas/tests/strings/test_find_replace.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/tests/strings/test_find_replace.py]: sure, updated

**Comment by jorisvandenbossche** [comment]
I think moving the shared `_str_find` method to the mixin would be best, if possible: that's consistent with how we are sharing code in other PRs now, and that also ensures that we don't have to care about the older backport, because this PR will simply put the fixed `_str_find` code in a new place, ensuring we get the older fix with it, also for StringDtype.

File: pandas/core/arrays/_arrow_string_mixins.py [hunk 2]
```diff
@@ -23,6 +26,13 @@ class ArrowStringArrayMixin:
     def __init__(self, *args, **kwargs) -> None:
         raise NotImplementedError
 
+    def _convert_int_dtype(self, result):
```
Comment (jorisvandenbossche) [inline pandas/core/arrays/_arrow_string_mixins.py:29]: Following https://github.com/pandas-dev/pandas/pull/59568/files#r1733236840, maybe `_convert_integer_result` to use a consistent pattern?

Comment (jbrockmendel) [reply pandas/core/arrays/_arrow_string_mixins.py:29]: how about a quick (and hopefully quickly-merged) PR to change the names on both this and _result_converter to something consistent?

File: pandas/core/arrays/string_arrow.py [hunk 3]
```diff
@@ -475,18 +475,14 @@ def _str_count(self, pat: str, flags: int = 0):
...
-        elif start == 0 and end is None:
-            slices = self._pa_array
-            result = pc.find_substring(slices, sub)
-        else:
+        if (
+            pa_version_under13p0
+            and not (start != 0 and end is not None)
+            and not (start == 0 and end is None)
+        ):
+            # https://github.com/pandas-dev/pandas/pull/59562/files#r1725688888
```
Comment (jorisvandenbossche) [inline pandas/core/arrays/string_arrow.py:483]: This link points to this PR, but it doesn't seem to work to actually link to a specific comment

Comment (jorisvandenbossche) [reply pandas/core/arrays/string_arrow.py:483]: Ah, I suppose it points to https://github.com/pandas-dev/pandas/pull/59562#discussion_r1725688888

Now, can't we move that into the mixin as well? (avoiding this override) This is something that was just buggy in pyarrow before that version AFAIU, so I think there is no harm in also doing object-dtype fallback for ArrowDtype, since it otherwise just errors wrongly

Comment (jbrockmendel) [reply pandas/core/arrays/string_arrow.py:483]: huh i couldve sworn i copy/pasted it from somewhere, but now i cant find it.  will update

File: pandas/core/arrays/_arrow_string_mixins.py [hunk 4]
```diff
@@ -23,6 +26,13 @@ class ArrowStringArrayMixin:
...
+        # Convert an int-dtype arrow result to an appropriate output type.
+        raise NotImplementedError
+
+    def _apply_elementwise(self, func: Callable) -> list[list[Any]]:
```
Comment (jorisvandenbossche) [inline pandas/core/arrays/_arrow_string_mixins.py:33]: This needs to be implemented for ArrowStringArray as well then?

File: pandas/core/arrays/_arrow_string_mixins.py (hunk 2 above)
Comment (jorisvandenbossche) [reply pandas/core/arrays/_arrow_string_mixins.py:29]: Yes, that would be my preference as well, then in all related PRs we can directly use the new names.

You had a branch starting that?

Comment (jbrockmendel) [reply pandas/core/arrays/_arrow_string_mixins.py:29]: will open a PR shortly

File: pandas/core/arrays/_arrow_string_mixins.py (hunk 4 above)
Comment (jbrockmendel) [reply pandas/core/arrays/_arrow_string_mixins.py:33]: this is only used for the ArrowEA version.  The ArrowStringArray goes through _str_map, which ArrowEA doesn't have.  eventually id like to align the names, but there are too many branches/PRs as it is.

Comment (jorisvandenbossche) [reply pandas/core/arrays/_arrow_string_mixins.py:33]: Not sure what I am missing, but `_apply_elementwise` is called from the now-shared `_str_find` method just below, and so I would think that you can also get there from `ArrowStringArray._str_find` ?

Comment (jbrockmendel) [reply pandas/core/arrays/_arrow_string_mixins.py:33]: yep my bad.  ArrowStringArray inherits ArrowEA so gets its apply_elementwise from there.  putting it here just prevents mypy from complaining

**Comment by jbrockmendel** [comment]
I think comments have been addressed here

File: pandas/core/arrays/string_arrow.py (hunk 3 above)
Comment (jorisvandenbossche) [reply pandas/core/arrays/string_arrow.py]: Can you check my question in my second comment above?

Comment (jbrockmendel) [reply pandas/core/arrays/string_arrow.py]: About moving the implementation to the mixin?  This PR now does that.

Comment (jorisvandenbossche) [reply pandas/core/arrays/string_arrow.py]: It's about the `if` block that this thread is commenting on, that is not in the mixin and my question is whether it shouldn't be moved as well? (it's a bug that affects ArrowEA as well AFAIU)

Comment (jbrockmendel) [reply pandas/core/arrays/string_arrow.py]: thanks for clarifying.  will give it a try.

File: pandas/core/arrays/string_arrow.py [hunk 5]
```diff
@@ -437,18 +437,14 @@ def _str_count(self, pat: str, flags: int = 0):
...
-        else:
+        if (
+            pa_version_under13p0
+            and not (start != 0 and end is not None)
+            and not (start == 0 and end is None)
+        ):
+            # GH#59562
             return super()._str_find(sub, start, end)
-        return self._convert_int_result(result)
+        return ArrowStringArrayMixin._str_find(self, sub, start, end)
```
Comment (jorisvandenbossche) [inline pandas/core/arrays/string_arrow.py:447]: Now that this special case is moved in the mixin method, I would expect this can be removed entirely? (and replaced with a `_str_find = ArrowStringArrayMixin._str_find`)

Comment (jorisvandenbossche) [reply pandas/core/arrays/string_arrow.py:447]: Ah, through `_str_map` using `lib.map_infer_mask`, I suppose. But if there is a cython implementation that is presumably faster, shouldn't we use that for the ArrowDtype as well? 
I saw that in the center PR at https://github.com/pandas-dev/pandas/pull/59624/files#diff-ca6e5560b2fc1721e129b85f10882df8a1f20b5f1ef4dff547170fa35898dfa6R62 you didn't use `_apply_elementwise` but also explicitly went through object dtype. That's for the same reason? Can we use the same pattern?

**Comment by jbrockmendel** [comment]
Apply element wise is the existing pattern the ArrowEA uses instead of
strmap. At some point I’d like to align/optimize it, but am not there yet.

**Comment by jorisvandenbossche** [comment]
Yes, but I am talking about two open PRs. Can't we already align in that? (I am not speaking about other code in `arrow/array.py`, but only about things we now refactor and move to `_arrow_string_mixins.py`

File: pandas/core/arrays/string_arrow.py (hunk 5 above)
Comment (jbrockmendel) [reply pandas/core/arrays/string_arrow.py]: Sure, changed.

Comment (jbrockmendel) [reply pandas/core/arrays/string_arrow.py:457]: looks like doing this broke the min_versions build, so reverted

File: pandas/tests/extension/test_arrow.py [hunk 6]
```diff
@@ -1990,11 +1977,15 @@ def test_str_find_e2e(start, end, sub):
...
     )
-    object_series = s.astype(pd.StringDtype())
+    object_series = s.astype(pd.StringDtype(storage="python"))
     result = s.str.find(sub, start, end)
     expected = object_series.str.find(sub, start, end).astype(result.dtype)
     tm.assert_series_equal(result, expected)
 
+    arrow_str_series = s.astype(pd.StringDtype(storage="pyarrow"))
+    result2 = arrow_str_series.str.find(sub, start, end).astype(result.dtype)
+    tm.assert_series_equal(result2, expected)
```
Comment (jorisvandenbossche) [inline pandas/tests/extension/test_arrow.py:1987]: For future PRs, we should add such tests to `pandas/tests/strings`, I think (because now it is testing StringDtype in tests specifically for ArrowDtype ..)

**Review by jorisvandenbossche — APPROVED** [review]

**Comment by jorisvandenbossche** [comment]
Thanks!
